eth2_testgen/bls/bls.py

```diff
@@ -0,0 +1,268 @@
+from typing import (  # noqa: F401
+    Dict,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+from eth_utils import (
+    big_endian_to_int,
+    ValidationError,
+)
+
+from py_ecc.optimized_bls12_381 import (  # NOQA
+    G1,
+    G2,
+    Z1,
+    Z2,
+    neg,
+    add,
+    multiply,
+    FQ,
+    FQ2,
+    FQ12,
+    FQP,
+    pairing,
+    normalize,
+    field_modulus as q,
+    b,
+    b2,
+    is_on_curve,
+    curve_order,
+    final_exponentiate
+)
+from hash import hash_eth2
+
+from eth2_typing import (
+    BLSPubkey,
+    BLSSignature,
+)
+
+G2_cofactor = 305502333931268344200999753193121504214466019254188142667664032982267604182971884026507427359259977847832272839041616661285803823378372096355777062779109  # noqa: E501
+FQ2_order = q ** 2 - 1
+eighth_roots_of_unity = [
+    FQ2([1, 1]) ** ((FQ2_order * k) // 8)
+    for k in range(8)
+]
+
+
+#
+# Helpers
+#
+def FQP_point_to_FQ2_point(pt: Tuple[FQP, FQP, FQP]) -> Tuple[FQ2, FQ2, FQ2]:
+    """
+    Transform FQP to FQ2 for type hinting.
+    """
+    return (
+        FQ2(pt[0].coeffs),
+        FQ2(pt[1].coeffs),
+        FQ2(pt[2].coeffs),
+    )
+
+
+def modular_squareroot(value: FQ2) -> FQP:
+    """
+    ``modular_squareroot(x)`` returns the value ``y`` such that ``y**2 % q == x``,
+    and None if this is not possible. In cases where there are two solutions,
+    the value with higher imaginary component is favored;
+    if both solutions have equal imaginary component the value with higher real
+    component is favored.
+    """
+    candidate_squareroot = value ** ((FQ2_order + 8) // 16)
+    check = candidate_squareroot ** 2 / value
+    if check in eighth_roots_of_unity[::2]:
+        x1 = candidate_squareroot / eighth_roots_of_unity[eighth_roots_of_unity.index(check) // 2]
+        x2 = FQ2([-x1.coeffs[0], -x1.coeffs[1]])  # x2 = -x1
```

Comment:
Due to https://github.com/ethereum/py_ecc/issues/47 (and the extensive discussion in https://github.com/ethereum/eth2.0-specs/issues/508, this line should be replaced by

```python
x2 = FQ2([-x1.coeffs[0] % q, -x1.coeffs[1] % q])  # x2 = -x1
```